modules/spring-config/src/main/java/org/mule/runtime/config/internal/DependencyGraphLifecycleObjectSorter.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.config.internal;
+
+import static java.util.Arrays.asList;
+import static java.util.Arrays.stream;
+import static java.util.Collections.reverse;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+import org.mule.runtime.config.internal.registry.AbstractSpringRegistry;
+import org.mule.runtime.config.internal.resolvers.DependencyGraphBeanDependencyResolver;
+import org.mule.runtime.core.internal.lifecycle.phases.DefaultLifecycleObjectSorter;
+import org.mule.runtime.core.internal.lifecycle.phases.LifecycleObjectSorter;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+
+import org.jgrapht.alg.cycle.CycleDetector;
+import org.jgrapht.graph.DefaultDirectedGraph;
+import org.jgrapht.graph.DefaultEdge;
+import org.jgrapht.traverse.TopologicalOrderIterator;
+import org.springframework.beans.factory.config.BeanDefinition;
+
+/**
+ * Specialization of {@link DefaultLifecycleObjectSorter} which uses an {@link AbstractSpringRegistry} to not only consider the
+ * provided objects but the beans on which that object depends on. This is accomplished by introspecting the
+ * {@link BeanDefinition} that was derived from the {@link Inject} annotations. This sorter was introduced to replace
+ * SpringLifecycleObjectSorter.
+ *
+ * @since 4.5.0
+ */
+public class DependencyGraphLifecycleObjectSorter implements LifecycleObjectSorter {
+
+  private List<DefaultDirectedGraph<BeanWrapper, DefaultEdge>> dependencyGraphs;
+  private DependencyGraphBeanDependencyResolver resolver;
+  protected Class<?>[] orderedLifecycleTypes;
+  private Map<String, Integer> lifecycleObjectNameOrderMap;
+
+  public DependencyGraphLifecycleObjectSorter(DependencyGraphBeanDependencyResolver resolver, Class<?>[] orderedLifecycleTypes) {
+    this.dependencyGraphs = new ArrayList<>();
+    this.resolver = resolver;
+    this.orderedLifecycleTypes = orderedLifecycleTypes;
```

Comment:
QQ: Do you want to copy the array or just hold a reference? Take into account that the parameter is mutable and you are using the field in other methods of this class

Reply:
Just holding a reference would be fine. I will mark it as final.